straxen/plugins/event_shadow.py

```diff
@@ -4,67 +4,48 @@
+    depends_on = ('event_basics', 'peak_basics', 'peak_shadow')
     provides = 'event_shadow'
     save_when = strax.SaveWhen.EXPLICIT
-    
 
     def infer_dtype(self):
-        dtype = [
-                ('pre_s2_area', np.float32,'previous s2 area [PE]'),
-                 ('shadow_dt', np.int64,'time diffrence to the previous s2 [ns]'),
-                 ('shadow', np.float32,'previous s2 shadow [PE/ns]'),
-                ('pre_s2_x', np.float32,'x of previous s2 peak causing shadow [cm]'),
-                 ('pre_s2_y', np.float32,'y of previous s2 peak causing shadow [cm]'),
-                 ('shadow_distance', np.float32,'distance to the previous s2 peak causing the max shadow [cm]')           
-                ]
+        dtype = [('shadow', np.float32, 'previous s2 shadow [PE/ns]'),
+                 ('pre_s2_area', np.float32, 'previous s2 area [PE]'),
+                 ('shadow_dt', np.int64, 'time diffrence to the previous s2 [ns]'),
+                 ('shadow_index', np.int32, 'max shadow peak index in event'),
+                 ('pre_s2_x', np.float32, 'x of previous s2 peak causing shadow [cm]'),
+                 ('pre_s2_y', np.float32, 'y of previous s2 peak causing shadow [cm]'),
+                 ('shadow_distance', np.float32, 'distance to the s2 peak with max shadow [cm]')]
         dtype += strax.time_fields
         return dtype
-    
+
     def compute(self, events, peaks):
-        roi_dt = np.dtype([(('back in time', 'time'), int),
-                                         (('till it begin','endtime'), int)])
-        roi = np.zeros(len(events), dtype=roi_dt)   
-        n_seconds = self.config['time_window_backward']
-        roi['time'] = events['time'] - n_seconds
-        roi['endtime'] = events['time']
-        mask_s2 = peaks['type'] == 2
-        mask_s2  &= peaks['area'] > self.config['pre_s2_area_threshold']   
-        split_peaks = strax.split_touching_windows(peaks[mask_s2], roi)
-        res = np.zeros(len(events), self.dtype)       
-        compute_shadow(events, split_peaks, res)
+        split_peaks = strax.split_by_containment(peaks, events)
+        res = np.zeros(len(events), self.dtype)
+        # Initialization
+        res['shadow_index'] = -1
+        res['pre_s2_x'] = np.nan
+        res['pre_s2_y'] = np.nan
+        for event_i, sp in enumerate(split_peaks):
+            if (sp['type'] == 2).sum() > 0:
+                # Define event shadow as the first S2 peak shadow
+                first_s2_index = np.argwhere(sp['type'] == 2)[0]
+                res['shadow_index'][event_i] = first_s2_index
+                res['shadow'][event_i] = sp['shadow'][first_s2_index]
+                res['pre_s2_area'][event_i] = sp['pre_s2_area'][first_s2_index]
+                res['shadow_dt'][event_i] = sp['shadow_dt'][first_s2_index]
+                res['pre_s2_x'][event_i] = sp['pre_s2_x'][first_s2_index]
+                res['pre_s2_y'][event_i] = sp['pre_s2_y'][first_s2_index]
```

Comment:
This part I do not understand. What is the motivation for having the first peaks shadow to be the one of the event? Because it is by definition of the shadow the largest one? But in this way you can also have the event itself casting a shadow on itself is not it? Should not it be therefore the first shadow for which `shadow_dt` corresponds to a time outside of the event?
In addition I do not understand why you are only looking for S2s. In principle also a S1 can be in the shadow of an S2 is not it?

Reply:
1. The reason we deliberately choose the first S2's shadow as the event's shadow is exactly as you say. I will check whether the performance of shadow variables changes after I change the peaks from the first to the peak with a `shadow_dt` out of the event. 
2. For the second question I think @chnlkx will have some comments?

Reply:
Yes, we believe large S1s might also cause isolated events, and possibly be different from the S2 shadow. We do plan to investigate the S1 shadow, but in the meantime, we would carry on the AC BG analysis with only S2 shadow and shadow distance first.

Reply:
Sorry, I did not mean that an S1 casts a shadow, but that an S1 is inside an S2 shadow.

Reply:
Yes, you are right. But most events have S2 but some events have no S1. Since peak-level shadow cut will suppress isoS1s, it is not necessary to define the shadow of an event as the shadow of any S1 peak in the event.

Reply:
Additionally, considering max drift time is small, the time window is almost the same for each peak in an event. So the first peak will always be the one with the largest shadow(have checked, for more than 99% of events, the first shadow is the largest). So I just defined shadow as the first S2 peak's shadow.

Reply:
Okay yes this make sense. Thanks for the clarification. Nice work.